shinylive/_url.py

```diff
@@ -0,0 +1,217 @@
+from __future__ import annotations
+
+import base64
+import json
+import os
+import re
+from pathlib import Path
+from typing import List, Literal, Optional, TypedDict, cast
+
+
+class FileContentJson(TypedDict):
+    name: str
+    content: str
+    type: Literal["text", "binary"]
+
+
+def encode_shinylive_url(
+    app: str | Path,
+    files: Optional[tuple[str | Path, ...]] = None,
+    mode: str = "editor",
+    language: Optional[str] = None,
+    header: bool = True,
+) -> str:
+    """
+    Generate a URL for a [ShinyLive application](https://shinylive.io).
+
+    Parameters
+    ----------
+    app
+        The main app file of the ShinyLive application. This file should be a Python
+        `app.py` or an R `app.R`, `ui.R`, or `server.R` file. This file will be renamed
+         `app.py` or `app.R` for shinylive, unless it's named `ui.R` or `server.R`.
+    files
+        A tuple of file or directory paths to include in the application. On shinylive,
+        these files will be stored relative to the main `app` file. If an entry in files
+        is a directory, then all files in that directory will be included, recursively.
+    mode
+        The mode of the application. Defaults to "editor".
+    language
+        The language of the application. Defaults to None.
+    header
+        Whether to include a header. Defaults to True.
+
+    Returns
+    -------
+        The generated URL for the ShinyLive application.
+    """
+
+    if language is None:
+        language = detect_app_language(app)
+
+    # if app has a newline, then it's app content, not a path
+    if isinstance(app, str) and "\n" in app:
+        app_path = ""
+        root_dir = Path(".")
+        file_bundle = [
+            {
+                "name": f"app.{'py' if language == 'py' else 'R'}",
+                "content": app,
+                "type": "text",
+            }
+        ]
+    else:
+        app_path = Path(app)
+        root_dir = app_path.parent
+        file_bundle = [read_file(app, root_dir)]
+
+    if files is not None:
+        file_list: list[str | Path] = []
+
+        for file in files:
+            if Path(file).is_dir():
+                file_list.extend(listdir_recursive(file))
+            else:
+                file_list.append(file)
+
+        file_bundle = file_bundle + [
+            read_file(file, root_dir) for file in file_list if Path(file) != app_path
+        ]
+
+    if language in ["py", "python"]:
+        language = "py"
+    elif language in ["r", "R"]:
+        language = "r"
+    else:
+        raise ValueError(
+            f"Language '{language}' is not supported. Please specify one of 'py' or 'r'."
+        )
+
+    # if first file is not named either `ui.R` or `server.R`, then make it app.{language}
+    if file_bundle[0]["name"] not in ["ui.R", "server.R"]:
+        file_bundle[0]["name"] = f"app.{'py' if language == 'py' else 'R'}"
```

Comment:
As it is currently written, I'm requiring that the primary app file be given to the `app` argument and then ensuring that it's the first item in the `file_bundle`.

The two lines highlighted in this comment ensure that, if the primary app file isn't `ui.R` or `server.R`, then it will be renamed `app.R` or `app.py` according to `language`. This setup is driven by shinylive requiring `app.py` or app/ui/server in R and the fact that locally we often store the primary app in a file that _isn't_ named `app.py`.

The current worst case scenario is that `ui.R` or `server.R` come first and the other file in the pair are later in the tab order, but if it really matters to someone they can make the second file the first item in `files`.